File: go/S0006.go
package main

import "fmt"
import "strings"

func convert(s string, numRows int) string {
    if numRows == 1 {
        return s
    }
    resultArr := make([]string, numRows)
    n := 2 * numRows - 2
    for i, char := range s {
        x := i % n
        resultArr[min(x, n - x)] += string(char)
    }
    return strings.Join(resultArr, "")
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func main() {
    fmt.Println(convert("LEETCODEISHIRING", 3))
}

File: ts/S0006.ts
function convert(s: string, numRows: number): string {
    if (numRows === 1) return s;

    let diff: 1 | -1 = -1, line = 0;
    let resultArr = Array(numRows).fill('');
    
    for (let i = 0; i < s.length; i++) {
        resultArr[line] += s[i];
        diff *= line % (numRows - 1) ? 1 : -1;
        line += diff;
    }
    return resultArr.reduce((str, item) => str += item, '');
};

console.log(convert('LEETCODEISHIRING', 3));
